<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Natural Language BioMart Queries • wizrd</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Natural Language BioMart Queries">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">wizrd</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/wizrd.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lawremi/wizrd" class="external-link">
    <span class="fa fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Natural Language BioMart Queries</h1>
                        <h4 data-toc-skip class="author">Michael
Lawrence</h4>
            
            <h4 data-toc-skip class="date">2025-07-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lawremi/wizrd/blob/main/vignettes/biomart-mcp.Rmd" class="external-link"><code>vignettes/biomart-mcp.Rmd</code></a></small>
      <div class="hidden name"><code>biomart-mcp.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette demonstrates how to use the wizrd agent interface with
the BioMart MCP server to perform natural language queries against
Ensembl and other BioMart resources. Unlike traditional biomaRt
workflows, you do not need to know the exact dataset, attribute, or
filter names—just ask your question in plain English and get a
data.frame result.</p>
<div class="section level3">
<h3 id="important-note-pybiomart-limitations">Important Note: pybiomart Limitations<a class="anchor" aria-label="anchor" href="#important-note-pybiomart-limitations"></a>
</h3>
<p>The BioMart MCP server uses the <code>pybiomart</code> Python
package, which relies on an older XML-based approach to BioMart. This
limits its functionality compared to the <code>biomaRt</code> R package,
which uses direct API access. Some queries that work with biomaRt may
not be possible with the current MCP server.</p>
</div>
</div>
<div class="section level2">
<h2 id="starting-the-biomart-mcp-server">Starting the BioMart MCP Server<a class="anchor" aria-label="anchor" href="#starting-the-biomart-mcp-server"></a>
</h2>
<p>We begin by downloading the BioMart MCP server script to a temporary
directory and launching it with all dependencies using <code>uvx</code>
and <code><a href="../reference/mcp.html">start_mcp()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lawremi/wizrd" class="external-link">wizrd</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">biomart_mcp_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcp"</span>, <span class="st">"biomart-mcp.py"</span>, package <span class="op">=</span> <span class="st">"wizrd"</span><span class="op">)</span></span>
<span><span class="va">server</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mcp.html">start_mcp</a></span><span class="op">(</span><span class="va">biomart_mcp_path</span><span class="op">)</span></span></code></pre></div>
<p>Connect to the BioMart MCP server.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">session</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mcp.html">connect_mcp</a></span><span class="op">(</span><span class="va">server</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, you can connect to a hosted BioMart MCP server
using:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">session</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mcp.html">connect_mcp</a></span><span class="op">(</span><span class="st">"https://server.smithery.ai/@jzinno/biomart-mcp/mcp"</span><span class="op">)</span></span></code></pre></div>
<p>This option requires a free account on smithery.ai and uses
interactive authorization, so this vignette will use the local
server.</p>
</div>
<div class="section level2">
<h2 id="creating-a-natural-language-biomart-agent">Creating a Natural Language BioMart Agent<a class="anchor" aria-label="anchor" href="#creating-a-natural-language-biomart-agent"></a>
</h2>
<p>We create an agent, equip it with the BioMart MCP tools, and instruct
it to answer user queries using BioMart. All outputs are returned as
data.frames for easy downstream analysis.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biomart_tools</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mcp.html">tools</a></span><span class="op">(</span><span class="va">session</span><span class="op">)</span></span>
<span><span class="va">agent</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/model-constructors.html">openai_agent</a></span><span class="op">(</span><span class="st">"o4-mini"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/equip.html">equip</a></span><span class="op">(</span><span class="va">biomart_tools</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/instruct.html">instruct</a></span><span class="op">(</span></span>
<span>        <span class="st">"You are an expert in Ensembl gene annotation."</span>,</span>
<span>        <span class="st">"Answer user queries using BioMart,"</span>,</span>
<span>        <span class="st">"using only strings for filter values, with lists comma-separated."</span>,</span>
<span>        <span class="st">"Return results as a data.frame."</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/output_as.html">output_as</a></span><span class="op">(</span><span class="fu">S7</span><span class="fu">::</span><span class="va"><a href="https://rconsortium.github.io/S7/reference/base_s3_classes.html" class="external-link">class_data.frame</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-1-gene-coordinates-cf--biomart-section-3-1">Example 1: Gene Coordinates (cf. biomaRt Section 3.1)<a class="anchor" aria-label="anchor" href="#example-1-gene-coordinates-cf--biomart-section-3-1"></a>
</h2>
<p>In biomaRt, you must specify the dataset, attributes, and filters.
With wizrd, just ask:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coords</span> <span class="op">&lt;-</span> <span class="va">agent</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/output_as.html">output_as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span></span>
<span>        <span class="st">"What are the chromosomal coordinates for TP53 and BRCA2?"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">coords</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span>keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames            ranges strand</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;   [1]       17   7661779-7687546      -</span></span>
<span><span class="co">#&gt;   [2]       13 32315086-32400268      +</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>The above relies on the GenomicRanges package in order to return a
useful R data structure for further computations, but every query can
return a simple data.frame, as demonstrated in later examples.</p>
<div class="section level3">
<h3 id="biomart-equivalent">biomaRt equivalent<a class="anchor" aria-label="anchor" href="#biomart-equivalent"></a>
</h3>
<p>The biomaRt equivalent requires the user to learn how to use the
biomaRt package, to determine many different identifiers, and to exert
extra effort to convert the output to a GRanges.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Huber-group-EMBL/biomaRt" class="external-link">biomaRt</a></span><span class="op">)</span></span>
<span><span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html" class="external-link">useMart</a></span><span class="op">(</span><span class="st">"ensembl"</span>, dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html" class="external-link">getBM</a></span><span class="op">(</span></span>
<span>    attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"external_gene_name"</span>, <span class="st">"chromosome_name"</span>, <span class="st">"start_position"</span>,</span>
<span>        <span class="st">"end_position"</span>, <span class="st">"strand"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    filters <span class="op">=</span> <span class="st">"external_gene_name"</span>,</span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"BRCA2"</span><span class="op">)</span>,</span>
<span>    mart <span class="op">=</span> <span class="va">mart</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/transform.html" class="external-link">transform</a></span><span class="op">(</span>strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">strand</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span></span>
<span>        seqnames.field <span class="op">=</span> <span class="st">"chromosome_name"</span>,</span>
<span>        start.field <span class="op">=</span> <span class="st">"start_position"</span>,</span>
<span>        end.field <span class="op">=</span> <span class="st">"end_position"</span>,</span>
<span>        keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="example-2-go-term-query-cf--biomart-section-3-6">Example 2: GO Term Query (cf. biomaRt Section 3.6)<a class="anchor" aria-label="anchor" href="#example-2-go-term-query-cf--biomart-section-3-6"></a>
</h2>
<p>Retrieve all EntrezGene IDs and gene symbols for genes with MAP
kinase activity (<a href="GO:0004707" class="uri">GO:0004707</a>):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_genes</span> <span class="op">&lt;-</span> <span class="va">agent</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span></span>
<span>        <span class="st">"List all EntrezGene IDs and gene symbols with MAP kinase activity"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">go_genes</span></span>
<span><span class="co">#&gt;    entrezgene_id hgnc_symbol</span></span>
<span><span class="co">#&gt; 1           5596       MAPK4</span></span>
<span><span class="co">#&gt; 2           5594       MAPK1</span></span>
<span><span class="co">#&gt; 3           6300      MAPK12</span></span>
<span><span class="co">#&gt; 4           5600      MAPK11</span></span>
<span><span class="co">#&gt; 5           5595       MAPK3</span></span>
<span><span class="co">#&gt; 6         225689      MAPK15</span></span>
<span><span class="co">#&gt; 7           5598       MAPK7</span></span>
<span><span class="co">#&gt; 8           5602      MAPK10</span></span>
<span><span class="co">#&gt; 9           5603      MAPK13</span></span>
<span><span class="co">#&gt; 10          5609      MAP2K7</span></span>
<span><span class="co">#&gt; 11          5601       MAPK9</span></span>
<span><span class="co">#&gt; 12          5599       MAPK8</span></span>
<span><span class="co">#&gt; 13          5597       MAPK6</span></span>
<span><span class="co">#&gt; 14          6885      MAP3K7</span></span>
<span><span class="co">#&gt; 15         51701         NLK</span></span>
<span><span class="co">#&gt; 16          1432      MAPK14</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-3-genes-on-specific-chromosomes-cf--biomart-section-3-3">Example 3: Genes on Specific Chromosomes (cf. biomaRt Section
3.3)<a class="anchor" aria-label="anchor" href="#example-3-genes-on-specific-chromosomes-cf--biomart-section-3-3"></a>
</h2>
<p>Retrieve all HUGO gene symbols of genes located on chromosome 20 and
associated with a specific GO term:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chr_go_genes</span> <span class="op">&lt;-</span> <span class="va">agent</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>        <span class="st">"Retrieve all HUGO gene symbols of genes on chromosome 20"</span>,</span>
<span>        <span class="st">"associated with GO:0005515 (protein binding)"</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">chr_go_genes</span><span class="op">)</span></span>
<span><span class="co">#&gt;   external_gene_name</span></span>
<span><span class="co">#&gt; 1             ADRA1D</span></span>
<span><span class="co">#&gt; 2              RSPO4</span></span>
<span><span class="co">#&gt; 3            DEFB129</span></span>
<span><span class="co">#&gt; 4              SHLD1</span></span>
<span><span class="co">#&gt; 5             SLX4IP</span></span>
<span><span class="co">#&gt; 6               BMP2</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>With wizrd and the BioMart MCP server, you can perform complex
biological queries using natural language, without needing to know the
technical details of BioMart datasets, attributes, or filters. All
results are returned as data.frames, making downstream analysis in R
seamless and efficient.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Michael Lawrence.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
